switchapi/src/switch_packet.c

```diff
@@ -410,3 +594,415 @@ int start_switch_api_packet_driver()
+        if (handle_type == SWITCH_HANDLE_TYPE_LAG) {
+            lag_info = switch_api_lag_get_internal(rx_key->port_lag_handle);
+            if (!lag_info) {
+                SWITCH_API_ERROR("invalid lag handle");
+                return SWITCH_STATUS_INVALID_HANDLE;
+            }
+            ifindex = lag_info->ifindex;
+        } else {
+            port_info = switch_api_port_get_internal(rx_key->port_lag_handle);
+            if (!port_info) {
+                SWITCH_API_ERROR("invalid port handle");
+                return SWITCH_STATUS_INVALID_HANDLE;
+            }
+            ifindex = port_info->ifindex;
+        }
+        rx_entry.ifindex_valid = TRUE;
+    }
+
+    if (rx_key->handle_valid) {
+        bd_handle = rx_key->handle;
+        handle_type = switch_handle_get_type(rx_key->handle);
+        if (handle_type == SWITCH_HANDLE_TYPE_INTERFACE) {
+            intf_info = switch_api_interface_get(rx_key->handle);
+            if (!intf_info) {
+                SWITCH_API_ERROR("intf_handle %lx is invalid", rx_key->handle);
+                return SWITCH_STATUS_INVALID_HANDLE;
+            }
+            if (!SWITCH_INTF_IS_PORT_L3(intf_info)) {
+                SWITCH_API_ERROR("intf_handle %lx is not l3", rx_key->handle);
+                return SWITCH_STATUS_INVALID_HANDLE;
+            }
+            bd_handle = intf_info->bd_handle;
+        }
+
+        bd_info = switch_bd_get(bd_handle);
+        if (!bd_info) {
+            SWITCH_API_ERROR("bd derivation failed %lx", rx_key->handle);
+            return SWITCH_STATUS_INVALID_HANDLE;
+        }
+        rx_entry.bd_valid = TRUE;
+    }
+
+    hostif_info = switch_hostif_get(rx_action->hostif_handle);
+    if (!hostif_info) {
+        SWITCH_API_ERROR("invalid hostif handle");
+        return SWITCH_STATUS_INVALID_HANDLE;
+    }
+
+    rx_entry.bd = handle_to_id(bd_handle);
+    rx_entry.ifindex = ifindex;
+    rx_entry.port = handle_to_id(rx_key->port_handle);
+    rx_entry.port_valid = rx_key->port_valid;
+    rx_entry.reason_code = rx_key->reason_code;
+    rx_entry.reason_code_valid = rx_key->reason_code_valid;
+    rx_entry.priority = rx_key->priority;
+
+    rx_info = switch_malloc(sizeof(switch_packet_rx_info_t), 0x1);
+    if (!rx_info) {
+        SWITCH_API_ERROR("port %lx port_lag %lx handle %lx malloc failed",
+                          rx_key->port_handle,
+                          rx_key->port_lag_handle,
+                          rx_key->handle);
+        return SWITCH_STATUS_NO_MEMORY;
+    }
+
+    memset(rx_info, 0x0, sizeof(switch_packet_rx_info_t));
+    memcpy(&rx_info->rx_entry, &rx_entry, sizeof(rx_entry));
+    rx_info->vlan_id = rx_action->vlan_id;
+    rx_info->vlan_action = rx_action->vlan_action;
+    rx_info->intf_fd = hostif_info->intf_fd;
+
+    tommy_list_insert_head(&packet_rx_list, &(rx_info->node), rx_info);
+    tommy_list_sort(&packet_rx_list, switch_packet_rx_compare);
```

Comment:
Looks like tommy list does not support inserting element using external ordering (cmp) function. insert_head+sort is wasteful.